From 61361f1c9dfb6dd84d5f878d2aa966291e8cb3d4 Mon Sep 17 00:00:00 2001
From: Zhenghua Wang <zhenghua.wang@intel.com>
Date: Fri, 26 Jul 2013 14:10:47 +0800
Subject: [PORT FROM MAIN] toolbox: houdini hook rebase on MR1

BZ: 71657

toolbox houdini hook rebase on MR1

Change-Id: I5706175cded8251009628b57a3716b031e8cfa64
Orig-Change-Id: Iaf94de78eab7463f619796700bb3d86b8a0461b7
Signed-off-by: llin4 <ling.lin@intel.com>
Signed-off-by: Zhenghua Wang <zhenghua.wang@intel.com>
---
 init/init.c        |   14 ++++++++++++++
 toolbox/Android.mk |    5 +++++
 toolbox/cat.c      |   12 ++++++++++++
 3 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/init/init.c b/init/init.c
index 2ed6222..60a389f 100755
--- a/init/init.c
+++ b/init/init.c
@@ -31,6 +31,7 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/un.h>
+#include <sys/personality.h>
 
 #include <selinux/selinux.h>
 #include <selinux/label.h>
@@ -748,6 +749,18 @@ static int property_service_init_action(int nargs, char **args)
     return 0;
 }
 
+static int personality_init_action(int nargs, char **args)
+{
+    const char *pval;
+    pval = property_get("ro.config.personality");
+    if (pval && !strcmp(pval, "compat_layout")) {
+        int old_personality;
+        old_personality = personality((unsigned long)-1);
+        personality(old_personality | ADDR_COMPAT_LAYOUT);
+    }
+    return 0;
+}
+
 static int signal_init_action(int nargs, char **args)
 {
     signal_init();
@@ -969,6 +982,7 @@ int main(int argc, char **argv)
     }
 
     queue_builtin_action(property_service_init_action, "property_service_init");
+    queue_builtin_action(personality_init_action, "personality_init");
     queue_builtin_action(signal_init_action, "signal_init");
     queue_builtin_action(check_startup_action, "check_startup");
 
diff --git a/toolbox/Android.mk b/toolbox/Android.mk
index c764690..aaad5b5 100644
--- a/toolbox/Android.mk
+++ b/toolbox/Android.mk
@@ -93,6 +93,11 @@ LOCAL_SHARED_LIBRARIES := \
 	libusbhost \
 	libselinux
 
+ifeq ($(INTEL_HOUDINI), true)
+    LOCAL_CFLAGS += -DWITH_HOUDINI
+    LOCAL_STATIC_LIBRARIES += houdini_hook
+endif
+
 LOCAL_MODULE := toolbox
 
 # Including this will define $(intermediates).
diff --git a/toolbox/cat.c b/toolbox/cat.c
index 6ac31f8..c80cbd3 100644
--- a/toolbox/cat.c
+++ b/toolbox/cat.c
@@ -183,6 +183,10 @@ raw_cat(int rfd)
 	}
 }
 
+#ifdef WITH_HOUDINI
+extern int houdini_hook_open(const char *path, int flags, int mode);
+#endif
+
 static void
 raw_args(char **argv)
 {
@@ -196,7 +200,11 @@ raw_args(char **argv)
 				fd = fileno(stdin);
 			else if (fflag) {
 				struct stat st;
+#ifdef WITH_HOUDINI
+				fd = houdini_hook_open(*argv, O_RDONLY|O_NONBLOCK, 0);
+#else
 				fd = open(*argv, O_RDONLY|O_NONBLOCK, 0);
+#endif
 				if (fd < 0)
 					goto skip;
 
@@ -210,7 +218,11 @@ raw_args(char **argv)
 					goto skipnomsg;
 				}
 			}
+#ifdef WITH_HOUDINI
+			else if ((fd = houdini_hook_open(*argv, O_RDONLY, 0)) < 0) {
+#else
 			else if ((fd = open(*argv, O_RDONLY, 0)) < 0) {
+#endif
 skip:
 				perror(*argv);
 skipnomsg:
-- 
1.7.4.1

